chatsky/core/context.py

```diff
@@ -36,18 +39,12 @@
 logger = logging.getLogger(__name__)
 
 
-def get_last_index(dictionary: dict) -> int:
-    """
-    Obtain the last index from the `dictionary`.
-
-    :param dictionary: Dictionary with unsorted keys.
-    :return: Last index from the `dictionary`.
-    :raises ValueError: If the dictionary is empty.
-    """
-    if len(dictionary) == 0:
-        raise ValueError("Dictionary is empty.")
-    indices = list(dictionary)
-    return max(indices)
+"""
+class Turn(BaseModel):
+    label: Optional[NodeLabel2Type] = Field(default=None)
+    request: Optional[Message] = Field(default=None)
+    response: Optional[Message] = Field(default=None)
+"""
```

Comment:
Not as a class.
We could add a method that zips all the turn fields for convenience:
```python
async def turns(self, slice) -> Iterable[Label, Message, Message]:
    return zip(*asyncio.gather(
        self.labels.__getitem__(slice),
        self.requests.__getitem__(slice),
        self.responses.__getitem__(slice)
    ))
```

Reply:
This also won't work well if we've added a request but haven't added a response or a label yet.
I think the code should be modified to pass absolute turn ids instead of slices to the context dict get method.
Something like:
```python
async def turns(self, slice) -> Iterable[Label, Message, Message]:
    turn_ids = range(self.current_turn_id + 1)[slice]
    return zip(*asyncio.gather(
        self.labels.__getitem__(turn_ids),
        self.requests.__getitem__(turn_ids),
        self.responses.__getitem__(turn_ids)
    ))
```

Reply:
Should probably add a method that also accepts slices but returns None on missing keys instead of throwing.

Reply:
```python
    async def get(self, key: K | Iterable[K], default=None) -> V | Iterable[V]:
        try:
            return await self[key]
        except KeyError:
            if isinstance(key, Iterable):
                return [self._items.get(k, default) for k in key]
            else:
                return default
```

Reply:
Wrong commit.